### PR TITLE
[11.0][FIX/IMP] hr_timesheet_sheet

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Update of https://github.com/OCA/hr-timesheet/pull/188.

I don't like two main things about this module:

 - One is that if you modify some data in the "Details" tab, you must have to save and click F5 to be able to see the same change in the "summary" tab.

- The other is that if you are in a timesheet sheet, already created, and in the form, in the "summary" tab you fill a 0 cell with some value, happens that if you cancel the edit and don't save the form, the timesheet related to that cell is already created.

So, I made this refactor to solve this two things. Now, the behaviour is like this:

* If you modify the `Details` tab, automatically the `Summary` tab is updated.
But if you modify the `Summary` tab, you need to save in order to have the `Details` tab updated.

* If you modify the `Summary` tab, and you need to do a change in the `Details` tab, please save before.

### CODE REVIEW

**Comment by mreficent** [comment]
Ready to review. Please, do a testing review, don't get lost in the code changes.

**Comment by MiquelRForgeFlow** [comment]
Ready to review. Please, do a testing review, don't get lost in the code changes.

@astirpe @alexey-pelykh @cubells

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 1]
```diff
@@ -554,34 +635,32 @@ def onchange_unit_amount(self):
                 'title': _("Warning"),
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I think this warning is not needed anymore. In this case also the workaround (the stuff with `context.get('params', {})`) can be removed.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 2]
```diff
@@ -235,7 +237,8 @@ def _get_timesheet_sheet_lines_domain(self):
             ('date', '<=', self.date_end),
             ('date', '>=', self.date_start),
             ('employee_id', '=', self.employee_id.id),
-            ('company_id', '=', self.employee_id.company_id.id),
+            # ('company_id', '=', self.company_id.id),
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: You will get a conflict with https://github.com/OCA/hr-timesheet/pull/223, just to let you know. Maybe you can make this PR dependent on https://github.com/OCA/hr-timesheet/pull/223 and rebase onto its branch?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 3]
```diff
@@ -398,8 +421,8 @@ def button_add_line(self):
         for rec in self:
             if rec.state in ['new', 'draft']:
                 rec.add_line(rec.add_line_project_id, rec.add_line_task_id)
-                rec.add_line_task_id = False
-                rec.add_line_project_id = False
+                rec.write(
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Probably is better to reuse (cherry-pick?) the [solution adopted in V12](https://github.com/OCA/hr-timesheet/blob/12.0/hr_timesheet_sheet/models/hr_timesheet_sheet.py#L432-L438).

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 4]
```diff
@@ -102,17 +102,19 @@ def _default_employee(self):
...
             'draft': [('readonly', False)],
-        }
+        },
     )
     line_ids = fields.One2many(
         comodel_name='hr_timesheet.sheet.line',
+        inverse_name='sheet_id',
         compute='_compute_line_ids',
         string='Timesheets',
+        store=True,
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Does now `def _compute_line_ids()` need the `@api.depends()`?

File: hr_timesheet_sheet/models/account_analytic_line.py [hunk 5]
```diff
@@ -25,9 +25,13 @@ def _compute_sheet(self):
                 ('company_id', 'in', [timesheet.company_id.id, False]),
                 ('state', 'in', ['new', 'draft']),
             ], limit=1)
-            if timesheet.sheet_id != sheet:
+            if sheet and timesheet.sheet_id != sheet:
```
Comment (astirpe) [inline hr_timesheet_sheet/models/account_analytic_line.py]: Even when `sheet == False and timesheet.sheet_id != sheet` you should do `timesheet.sheet_id = sheet`.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 6]
```diff
@@ -571,24 +613,22 @@ def onchange_unit_amount(self):
...
+        self._update_line()
+
+    def _update_line(self):
+        sheet = self.sheet_id
+        if self and not self.ids:
+            line = sheet.line_ids.filtered(
+                lambda l: l.date == self.date
+                and l.project_id.id == self.project_id.id
+                and l.task_id.id == self.task_id.id
+                and l.id != self.id
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:663]: `if not self.ids` ... and then ... `l.id != self.id` :thinking:

**Review by astirpe — CHANGES_REQUESTED** [review]
Tested quickly but I need more time to test it. Will do next week.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I would like you to test a bit more to be sure about this. Remember you provided tests 11 and 12, and both still work fine.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 2 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: It's fine to me. Find someone to merge #223 asap :dancing_women:

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: ok

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I was not able to reproduce the warning anymore. I will continue to test of course but if you also cannot reproduce it I would just remove it. Your solution is obsoleting this warning, I'm quite sure about this.

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 5 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/account_analytic_line.py]: Are you sure? Maybe you are having `timesheet.sheet_id=NewId...` I will try to check...

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 6 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:663]: hehe this case is when you have `self.id = NewId...`, so self is True, but self.ids is [], and in this filter you want to exclude this id, and find an already existing id. I have proved that while debugging, that you may have an existing sheet, but when you fill a cell, you are not editing the corresponding line but you are creating a new line. And when you have two lines in same cell problems may arise, so with this method I am assuring you only have a line per cell.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: good. And why the related tests are still working?

Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: The tests are triggering the workaround  and checking that this workaround works. But since the workaround cannot be triggered by human interaction anymore, this code and its related tests are now obsolete and can be removed.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 6 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:663]: Would be more clear written this way:

```python
if isinstance(self.id, models.NewId):```

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 5 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/account_analytic_line.py]: This method is invoked as `res._compute_sheet()` where `res` is the result of a create, so `sheet` is always defined.

Comment (astirpe) [reply hr_timesheet_sheet/models/account_analytic_line.py]: Sorry, wrong message above :rofl: I will delete it and rewrite it next week :smile:

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: ah, ok hahaha I didn't check that they were triggering specifically the workaround, so I can remove test 11, right? and the test 12?

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 5 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/account_analytic_line.py]: I found a case where sheet was False because the search didn't found a sheet. I will check why/when/how it happens.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 6 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:663]: niiiice

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 5 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/account_analytic_line.py]: Exactly! When the search does not find any sheet, `timesheet.sheet_id = sheet` should be done as well.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 2 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: rebased and comments attended =)

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: You could keep test 12, just updating [this line](https://github.com/OCA/hr-timesheet/blob/11.0/hr_timesheet_sheet/tests/test_hr_timesheet_sheet.py#L581) and removing [this line](https://github.com/OCA/hr-timesheet/blob/11.0/hr_timesheet_sheet/tests/test_hr_timesheet_sheet.py#L607).

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 2 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Merged!

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I think there is no need to update/remove those lines :S The onchange always run when modifying a cell. The tests are still green :)

**Review by astirpe — CHANGES_REQUESTED** [review]
After another quick test session:

field `line_ids` is now stored. This causes the following issue:

**Issue 1**

1. Create a timesheet sheet. Save.
2. Go to "Timesheet"->"All timesheet" and change one value (eg.: the time `unit_amount`) of one timesheet.
3. Open back the timesheet sheet: the Summary tab is not updated with the new `unit_amount` value. (not ok)
4. The Summary tab is not anymore consistent with the Details tab. (not ok)


_______



Model '`hr_timesheet.sheet.line`' is a transient model. It means that the records will be periodically cleaned by a cron job. This causes the following issue:

**Issue 2**

1. Create a timesheet sheet. Save.
2. If you do not want to wait that the cron job removes the sheet lines (it would take a while), simply start the Odoo server passing parameter `--osv-memory-age-limit=0.00001` (or any very low value). Then go to "Settings"->"Automation"->"Scheduled actions". Open "Base: Auto-vacuum internal data". Click on "Run Manually".
3. Open back the timesheet sheet: the Summary tab says: "Sorry no matrix data to display." (not ok)
4. The Summary tab is not anymore consistent with the Details tab. (not ok)

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 7]
```diff
@@ -320,9 +338,13 @@ def create(self, vals):
                 raise UserError(
                     _('In order to create a sheet for this employee, '
                       'you must link him/her to an user.'))
+        line_ids = self.env['hr_timesheet.sheet.line']
+        if 'line_ids' in vals:
+            for val in vals['line_ids']:
```
Comment (nikul-serpentcs) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: @mreficent Here ` for val in vals.get('line_ids', False):`
No need to check `if 'line_ids' in vals:`

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 8]
```diff
@@ -481,21 +516,71 @@ def clean_timesheets(self, timesheets):
...
+            if len(new_ts) > 1:
+                new_ts = new_ts.merge_timesheets()
+                sheet.with_context(in_write=True).write(
+                    {'timesheet_ids': [
+                        (6, 0, sheet.timesheet_ids.exists().ids)]})
+            if not diff_amount:
+                continue
+            if new_ts:
+                unit_amount = new_ts.unit_amount + diff_amount
+                if unit_amount:
```
Comment (nikul-serpentcs) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:581]: @mreficent here needed ` if unit_amount:`??

I think you can directly write `new_ts.write({'unit_amount': unit_amount or 0.0 })`

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
netpicking

**Comment by MiquelRForgeFlow** [comment]
@astirpe 
The first issue I think is solvable, by calling something in the write of the sheet lines. I will see hot to fix it.

The second issue it's not related with this PR, it has been always a issue from the beginning. It's solvable by calling the `_compute_line_ids` every time you try to edit the sheet (I don't know how) or by converting the transient model to a normal model.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 8 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:581]: yeah, but then you will have the `else` becoming `if not unit_amount:`, thus the check is done inevitably.

**Comment by astirpe** [comment]
@mreficent 
If the first issue requires extra development on the write, then my question is: why do we need this kind of changes? What's the benefit for the end user? Without this PR the summary tab is already updated when changing the details tab.

The second issue was not an issue before this PR, because `line_ids` was not stored and was always recomputed on the fly.
It's an issue now. Try to approve a timesheet sheet today and check tomorrow how is the timesheet sheet looking.

**Comment by MiquelRForgeFlow** [comment]
@astirpe 
to your first questions:
Without this PR, If you are in a timesheet sheet, already created, and in the form, in the "summary" tab you fill a 0 cell with some value, happens that if you cancel the edit and don't save the form, the timesheet related to that cell is already created and is not deleted.

About the second issue, yeah, you are right. I put the stored to solve another issue but it's true that being stored it's what avoids the recomputing on the fly. I will check if without the stored, I can solve the same issues I detected.

**Comment by MiquelRForgeFlow** [comment]
@astirpe I solved your last two issues :)

File: hr_timesheet_sheet/models/account_analytic_line.py [hunk 9]
```diff
@@ -38,15 +42,27 @@ def create(self, values):
     def write(self, values):
         self._check_state_on_write(values)
         res = super(AccountAnalyticLine, self).write(values)
-        vals_do_compute = ['date', 'employee_id', 'project_id', 'company_id']
+        vals_do_compute = [
+            'date', 'employee_id', 'project_id', 'task_id', 'company_id',
```
Comment (astirpe) [inline hr_timesheet_sheet/models/account_analytic_line.py]: Why do you change lines that are not related with this PR? Adding 'task_id' here is not necessary.

File: hr_timesheet_sheet/models/account_analytic_line.py [hunk 10]
```diff
@@ -38,15 +42,27 @@ def create(self, values):
...
+        ]
         if any(val in vals_do_compute for val in values):
             self._compute_sheet()
+        vals_do_lines_compute = [
+            'unit_amount', 'name',
```
Comment (astirpe) [inline hr_timesheet_sheet/models/account_analytic_line.py:49]: No need for 'name'. Instead 'date', 'employee_id', 'project_id', 'task_id', 'company_id' are needed.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 11]
```diff
@@ -148,6 +150,11 @@ def _default_employee(self):
         compute='_compute_total_time',
         store=True,
     )
+    _dummy_field = fields.Boolean(
+        compute='_compute_dummy_field',
+        store=False,
+        # hack: needed to force recompute "stored" lines
```
Comment (astirpe) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:156]: Instead of adding another hack, wasn't more logical to just avoid the model to be transient?

**Review by astirpe — CHANGES_REQUESTED** [review]
Not tested, just code review.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 11 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:156]: But if the model is not transient, it will occupy a lot of unnecessary data in the database.

Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:156]: What about `unlink()` ?

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 10 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/account_analytic_line.py:49]: Test before suggestions please.

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 9 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/account_analytic_line.py]: It's necessary because you may change the task of a timesheet without changing its project, and thus this kind of change should trigger also `_compute_sheet` method

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 11 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:156]: Anyway when I was working on this module I noticed that there were never leftover records for the lines, because the system automatically removed the orphan lines when recomputing. Not sure that now the behavior is different because of the `stored=True`. Would you give it a try? Maybe the `unlink()` is not even necessary.

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:156]: You will only unlink the deprecated lines, but you will not unlink the lines used for all existing sheets. And in a year of sheets for just an employee, already it's a lot of lines. I don't want to imagine how many lines this suppose for several employees :S

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 9 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/account_analytic_line.py]: Why you want to trigger `_compute_sheet` if you only change the task? The `sheet_id` stays the same in that case.

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 10 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/account_analytic_line.py:49]: Sure I will! In the meantime can you take in consideration my code reviews?

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 9 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/account_analytic_line.py]: 🤔 yeah, but the same I could say for `project_id`, right?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 11 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:156]: @mreficent sorry I don't understand your point. At the moment you create new lines you can also delete old obsolete lines. And this is not even necessary in case the system does it automatically for you.

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 9 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/account_analytic_line.py]: No. The project can belong to a different employee or different company.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 11 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:156]: What do you mean? I have a sheet for this week with 4 lines. Yeah, I can change this 4 rows for other different 4 rows. Unlink ones and create others. But, this sheet will always have 4 rows (and thus 4x7 lines occupying space in database).

File: hr_timesheet_sheet/models/account_analytic_line.py (hunk 9 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/account_analytic_line.py]: :thinking: oh, ok :+1:

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 11 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:156]: Yes, and what's the problem then? You made them stored, this is what are you proposing. Next time you open the timesheet sheet you will see the same lines again without need of recomputing.

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:156]: I prefer not storing the lines and recomputing them when opening a sheet form. I put the "store=True" knowing they are transient. The need of the "store=True" was to fix some bad behavior when you add rows in the summary using the add project button.

**Comment by MiquelRForgeFlow** [comment]
I will change the approach. No more magic. No more stored transients. We have found a nice and simple way to handle it.

**Comment by MiquelRForgeFlow** [comment]
Substituted (and fixed) by https://github.com/OCA/hr-timesheet/pull/237.
